dragonfly/implementations/psyllid_provider.py

```diff
@@ -0,0 +1,279 @@
+            self.freq_dict[channel]=None
+            return False
+        except Exception as e:
+            logger.error('Unexpected error in set_central_frequency')
+            raise 
+
+
+    def start_run(self, channel, duration, filename):
+        payload = {'duration':duration, 'filename':filename}
+        result = self.provider.cmd(self.queue_dict[channel], 'start-run', payload=payload)
+
+
+    def stop_run(self, channel):
+        result = self.provider.cmd(self.queue_dict[channel], 'stop-run')
+
+
+    @property
+    def number_of_channels(self):
+        active_channels = [i for i in self.freq_dict.keys() if self.freq_dict[i]!=None]
+        logger.info('Active channels are {}'.format(active_channels))
+        return len(active_channels)
+
+
+    @property
+    def active_channels(self):
+        active_channels = [i for i in self.status_value_dict.keys() if self.status_value_dict[i]==4]
+        return active_channels
+
+
+    def get_number_of_streams(self, channel):
+        stream_count = 0
+        for i in range(3):
+            try:
+                request = '.node-config.ch'+str(i)+'.strw'
+                result = self.provider.get(self.queue_dict[channel]+request)
+                stream_count += 1
+                self.mode_dict[channel]='streaming'
+            except core.exceptions.DriplineError:
+                try:
+                    request = '.node-config.ch'+str(i)+'.trw'
+                    result = self.provider.get(self.queue_dict[channel]+request)
+                    stream_count += 1
+                    self.mode_dict[channel]='triggered'
+                except core.exceptions.DriplineError:
+                    pass  
+        logger.info('Number of streams for channel {}: {}'.format(channel, stream_count))
+        return stream_count
+
+
+    def set_fmt_snr_threshold(self, snr, channel='a'):
+        if self.mode_dict[channel] == 'streaming':
+            logger.warning('Psyllid not in streaming mode')
+            return False
+        request = '.node-config.ch'+str(self.channel_dict[channel])+'.fmt.threshold-power-snr'
+        result = self.provider.set(self.queue_dict[channel]+request, snr)
+        self.deactivate(channel)
+        time.sleep(1)
+        self.activate(channel)
+        return True
+
+
+    def make_trigger_mask(self, channel='a', filename='~/fmt_mask.json'):
+        if self.mode_dict[channel] == 'streaming':
+            return False
+
+        self.request_status(channel)
+        if self.status_value_dict[channel] !=4:
+            self.activate(channel)
+
+        logger.info('freq-only')
+        request = 'run-daq-cmd.ch'+str(self.channel_dict[channel])+'.tfrr.freq-only'
+        result = self.provider.cmd(self.queue_dict[channel],request)
+
+        logger.info('update-mask')
+        request = 'run-daq-cmd.ch'+str(self.channel_dict[channel])+'.fmt.update-mask'
+        result = self.provider.cmd(self.queue_dict[channel],request)
+        time.sleep(1)
```

Comment:
Again, these sleeps feel very clunky. Hopefully it is the case that either: 1) the action is quick and the target is done when it sends its reply, or 2) the success Reply  only indicates that the request was received and valid but the action is ongoing and there is therefore a way to poll to see when it is done.

One of the bigger concerns here is that if we're in case 2 and not actually polling, then how do we know if there was an error in the last step of the "tfrr.freq-only" or "update-mask" commands? It isn't clear that they've had an opportunity to send a reply indicating an issue.